CHANGELOG.md

```diff
@@ -4,22 +4,22 @@
-- `Integers`, `Doubles`, `Strings`, `Logicals` and `Complexes`: `ALTREP` wrappers for vectors that allows for much larger vectors than native R [[#274]](https://github.com/extendr/extendr/pull/274), [[#284]](https://github.com/extendr/extendr/pull/284), [[#301]](https://github.com/extendr/extendr/pull/301), [[#338]](https://github.com/extendr/extendr/pull/338), [[#350]](https://github.com/extendr/extendr/pull/350)
+- `Function` type that wraps an R function, which can be invoked using the `call()` method. [[#188]](https://github.com/extendr/extendr/pull/188)
+- `pairlist!` macro for generating `Pairlist` objects, e.g. for use in function calls. [[#202]](https://github.com/extendr/extendr/pull/202)
+- `use_try_from` option for the `extendr` macro, which allows the use of any type that implements `TryInto<Robj>`/`TryFrom<Robj>`, e.g. `#[extendr(use_try_from = true)]`. [[#222]](https://github.com/extendr/extendr/pull/222)
+- Large Rust integer types (`u32`, `u64` and `i64`) can now be converted to R's `numeric` type, which can handle large integer values. [[#242]](https://github.com/extendr/extendr/pull/242)
+- `call!` macro, which can be used to call an R function whose name is provided as a string. [[#238]](https://github.com/extendr/extendr/pull/238)
+- `TryFrom<Robj>` for a large number of Rust types. [[#249]](https://github.com/extendr/extendr/pull/249), [[#258]](https://github.com/extendr/extendr/pull/258)
+- Support for `ALTREP`. [[#250]](https://github.com/extendr/extendr/pull/250)
+- `S4` struct, which wraps an S4 class in R. [[#268]](https://github.com/extendr/extendr/pull/268)
+- Implemented `TryFrom<&ArrayBase> for Robj`, allowing `extendr`-annotated functions to return Arrays from the `ndarray` crate and have them automatically converted to R arrays. Note: even if 1D arrays are returned they will not be returned as vectors.. [[#275]](https://github.com/extendr/extendr/pull/275)
+- `Rint`, `Rdouble`, `Rbool` and `Rcplx`: `NA`-aware wrappers classes for scalar elements of R vectors [[#274]](https://github.com/extendr/extendr/pull/274), [[#284]](https://github.com/extendr/extendr/pull/284), [[#301]](https://github.com/extendr/extendr/pull/301), [[#338]](https://github.com/extendr/extendr/pull/338), [[#350]](https://github.com/extendr/extendr/pull/350)
+- `Integers`, `Doubles`, `Strings`, `Logicals` and `Complexes`: `ALTREP` wrappers for R vectors that deref to slices of the above types (`RInt` etc). [[#274]](https://github.com/extendr/extendr/pull/274), [[#284]](https://github.com/extendr/extendr/pull/284), [[#301]](https://github.com/extendr/extendr/pull/301), [[#338]](https://github.com/extendr/extendr/pull/338), [[#350]](https://github.com/extendr/extendr/pull/350)
```

Comment:
These aren't ALTREPs as afar as I remember. These are just wrappers around R vectors. They do handle altreps through the standard C/R functions. There is a separate trait for ALTREP, which pretty much allows creating generators on R side by marshaling some state data and a generating function to the R side.
```suggestion
- `Integers`, `Doubles`, `Strings`, `Logicals` and `Complexes`: wrappers for R vectors (both regular and `ALTREP`) that deref to slices of the above types (`Rint` etc). [[#274]](https://github.com/extendr/extendr/pull/274), [[#284]](https://github.com/extendr/extendr/pull/284), [[#301]](https://github.com/extendr/extendr/pull/301), [[#338]](https://github.com/extendr/extendr/pull/338), [[#350]](https://github.com/extendr/extendr/pull/350)
```

Reply:
``` r
rextendr::rust_function(
    "fn deref_altrep(x : Integers, i : i32) -> extendr_api::scalar::Rint { (*x)[i as usize] }",
    patch.crates_io = list(`extendr-api` = list(git = "https://github.com/extendr/extendr"))
)
#> ℹ build directory: 'C:/Users/<...>/AppData/Local/Temp/RtmpyQgE7x/filec1bc5bb465c8'
#> ✔ Writing 'C:/Users/<...>/AppData/Local/Temp/RtmpyQgE7x/filec1bc5bb465c8/target/extendr_wrappers.R'

x <- 1L:1e7L
x |> lobstr::sxp() |> print()
#> [1:0x1fa53ee7250] <INTSXP[1e+07]> (altrep named:65535)

testthat::expect_equal(
    deref_altrep(x, 50L),
    51L
)
```

<sup>Created on 2022-05-29 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Reply:
What does this tell us?

Reply:
I am sorry I did not explain enough what I tried to check. Our PRs are a bit... mixed up, sometimes we push in several features in one PR. PRs you mention in this bullet point are all related to vector wrappers and their respective scalar types, which *are not exclusively ALTREP*. However, this PR https://github.com/extendr/extendr/pull/274 introduces not only `Rint`, but also mechanisms to produce custom ALTREP vectors, i.e. a way to convert rust data structure and rust function into an altrep class that R understands. I'd suggest to move this PR (or mention it again) at the `ALTREP` introduction bullet.

Here I suggest to write that theses types (`Integers` and friends) are wrappers for *all* R vectors, both regular and altrep. Otherwise it may seem that `Integers` are solely for `ALTREP`, which is not the case.
My example showed that one can pass an altrep integer vector to rust using `Integers` as wrapper type and perform deref to get one of the elements.

Here is the same function working identically with both altrep and non-altrep vectors 
``` r
rextendr::rust_source(
    code = "
    use extendr_api::prelude::*;
    #[extendr]
    fn deref_altrep(x : Integers, i : i32) -> Rint {
        (*x)[i as usize] 
    }
    ",
    patch.crates_io = list(`extendr-api` = list(git = "https://github.com/extendr/extendr"))
)
#> ℹ build directory: 'C:/Users/<...>/AppData/Local/Temp/RtmpOUbDTy/file86b87a9d7b76'
#> ✔ Writing 'C:/Users/<...>/AppData/Local/Temp/RtmpOUbDTy/file86b87a9d7b76/target/extendr_wrappers.R'

x <- 1L:1e7L
x |> lobstr::sxp() |> print()
#> [1:0x1c1f6d2feb0] <INTSXP[1e+07]> (altrep named:65535)

testthat::expect_equal(
    deref_altrep(x, 50L),
    51L
)
y <- c(1L, 2L, 3L, 4L, 5L, 6L, 7L)
y |> lobstr::sxp() |> print()
#> [1:0x1c1f6243a08] <INTSXP[7]> (named:4)

testthat::expect_equal(
    deref_altrep(y, 5L),
    6L
)
```

<sup>Created on 2022-05-29 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Reply:
Okay I've removed the altrep reference. I'm happy to reference the same PR multiple times, so I've put it under *both* altrep and under the vector type feature.